nvtabular/torch_dataloader.py

```diff
@@ -233,14 +219,36 @@ def __init__(
         self.itr = GPUFileIterator(path, **kwargs)
         self.batch_size = sub_batch_size
         self.num_chunks = len(self.itr.engine)
+        self.buffer = queue.Queue()
+        self.lock = threading.Lock()
 
     def __len__(self):
         return self.num_chunks
 
     def __iter__(self):
-        for chunk in self.itr:
-            for idx in range(0, len(chunk), self.batch_size):
-                yield chunk.iloc[idx : idx + self.batch_size]
+        self.itr = iter(self.itr)
+        self.load_chunk()
+        for x in range(self.num_chunks):
+            threading.Thread(target=self.load_chunk).start()
+            chunk = self.buffer.get()
+            yield from TensorItr(
+                chunk,
+                batch_size=self.batch_size,
+                cat_cols=self.cat_cols,
+                cont_cols=self.cont_cols,
+                label_cols=self.label_cols,
+            )
+
+    def load_chunk(self):
+        with self.lock:
+            try:
+                chunk = next(self.itr)
+            except StopIteration:
+                return
+            chunk = create_tensors_plain(
+                chunk, cat_cols=self.cat_cols, cont_cols=self.cont_cols, label_cols=self.label_cols
+            )
+            self.buffer.put(chunk)
```

Comment:
I don't like creating a new thread to read in a single chunk here. I also think this code is safe to call from multiple threads because of you are storing mutable state on the class with the itr/lock/buffer members (and lines like ```self.itr = iter(self.itr)``` probably mean that you can only iterate over this class once?).

How about something like this instead :

```suggestion
        buffer = queue.Queue(1)
        threading.Thread(target=self._load_chunks, args=(buffer,)).start()

        for _ in range(self.num_chunks):
            chunk = buffer.get()
            yield from TensorItr(
                chunk,
                batch_size=self.batch_size,
                cat_cols=self.cat_cols,
                cont_cols=self.cont_cols,
                label_cols=self.label_cols,
            )

    def _load_chunks(self, output):
        for chunk in self.itr:
            chunk = create_tensors_plain(
                chunk, cat_cols=self.cat_cols, cont_cols=self.cont_cols, label_cols=self.label_cols
            )
            output.put(chunk)
```

This only creates 1 background thread,  and removes the lock/buffer member variables - making the buffer a local variable, and using thread locking from the queue. To limit to 1 background read of a chunk at a time we cap the buffer size to 1.